### ISSUE TITLE
Nodes stop syncing with “Too far ahead for peer …” warnings

### ISSUE THREAD

**Issue by trietnguyen267** [paritytech/substrate#5550]
Issue occured when node is joining network or just randomly occurs when it sync with others. 

When `sync` module emitted warning `Too far ahead for peer`, the node drain ~4k blocks and re-import the chunk which eventually trigger the restriction again and again.

This behaviour made node unable to sync forward and only repeat the steps of draining big amount of already imported block then re-import them

Possibly related to the following change
https://github.com/paritytech/substrate/pull/4095/files

P/s: the nodes we're running included changes upto this commit hash: `44e2e1d953a4a78ca135ea5a7c9aa2c7df91f792`
```
4/6/2020, 10:58:22.657 PM	2020-04-06 22:58:22.657 main-tokio- TRACE sync Too far ahead for peer QmZuqux9gnymQeVozwKD1J72XSL6TF5K2D3a5ArDFQnkZ9 (59940)	
4/6/2020, 10:58:22.658 PM	2020-04-06 22:58:22.658 main-tokio- TRACE sync Too far ahead for peer QmNj9cdxndTgysJ6CJzHksHQVvUqjt1zMnCnLQeCuXtDUQ (59940)	
4/6/2020, 10:58:22.661 PM	2020-04-06 22:58:22.662 main-tokio- TRACE sync BlockResponse 87137 from QmTP4ywbknv5DEPXUKoZKD3wkxnBVpm48x5Qqcevay3D9m with 128 blocks (55972..56099)	
4/6/2020, 10:58:22.662 PM	2020-04-06 22:58:22.662 main-tokio- TRACE sync Drained 3968 blocks	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 main-tokio- TRACE sync Scheduling 3968 blocks for import	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.663 main-tokio- TRACE sync Accepted 3968 blocks (0x7bebd36ec8932b351924596b56978bd0e1dedba4c19885e4480f1d497f8346b4) with origin NetworkInitialSync	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 main-tokio- TRACE sync Too many blocks in the queue.	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Header 0xe6fb…11e1 has 2 logs	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Block already in chain 55974: 0xe6fb738cd4bf2b667df7fe81b4d5337f9403ad0ee65919a49bda7848f78011e1	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Header 0x4fef…8cd4 has 2 logs	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Block imported successfully Some(55972) (0x2441…c0e6)	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Block already in chain 55972: 0x2441fcdd87cf46cf2b61d0244ad594b19c84a7d3872e90e5ab53b4f161f9c0e6	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 main-tokio- TRACE sync Too many blocks in the queue.	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Starting import of 3968 blocks (55972..59939)	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 main-tokio- TRACE sync Too many blocks in the queue.	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Block already in chain 55973: 0x4fefac8aaa3cebb495a2796e41c0b89a582bbe58764da2eaa6148b2275c38cd4	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Block imported successfully Some(55973) (0x4fef…8cd4)	
4/6/2020, 10:58:22.663 PM	2020-04-06 22:58:22.664 import-queue-worker-0 TRACE sync Header 0x2441…c0e6 has 2 logs	
```

**Comment by bkchr** [comment]
CC @arkpar

**Comment by arkpar** [comment]
Mind posting the full log?
`Too far ahead for peer ` this simply means that we've don't want to request blocks to far ahead before pending request are complete. Which is normal.
`Drained 3968 blocks` That's also fine when a gap in the download map is filled. 
`Too many blocks in the queue` means download is paused until pending blocks are processed.
It is not clear why it re-download existing blocks though.

**Comment by trietnguyen267** [comment]
Hi @arkpar,
I'm attaching the full logs we have since node restart to a few secs after problem occurred.
[fullLog.zip](https://github.com/paritytech/substrate/files/4447234/fullLog.zip)

The re-download process seems to be occurred just 1 mins after node start syncing from the network.

**Comment by arkpar** [comment]
@trietnguyen267 Thanks, but it looks like this file only contains first 4000 lines of the log. It cuts only after a few seconds of run time.

**Comment by trietnguyen267** [comment]
Would you like the full log from where it start to the time it stuck? That would be millions of logs, will take sometime to collect all. How about logs at the first time this draining and re-import with same size occurred?

**Comment by arkpar** [comment]
60 seconds before it got stuck or re-import happened will do indeed

**Comment by trietnguyen267** [comment]
@arkpar this is the file contains first time the issue occurs and the repeating drain/import blocks
[full.log](https://github.com/paritytech/substrate/files/4473061/full.log)

**Comment by arkpar** [comment]
Should be fixed in #5604
